File: packages/native-helpers/windows-helper/src/Services/AudioService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Reflection;
using System.Runtime.InteropServices;
using System.Threading.Tasks;
using NAudio.Wave;
using NAudio.CoreAudioApi;

namespace WindowsHelper.Services
{
    public class AudioService
    {
        private WaveOutEvent? waveOut;
        private MMDeviceEnumerator? deviceEnumerator;
        private float originalVolume = 1.0f;
        private bool originalMuteState = false;

        // Preloaded audio data for faster playback
        private readonly Dictionary<string, byte[]> preloadedAudio = new();

        public event EventHandler<string>? SoundPlaybackCompleted;

        public AudioService()
        {
            try
            {
                deviceEnumerator = new MMDeviceEnumerator();

                // Preload audio files at startup for faster playback
                PreloadSound("rec-start");
                PreloadSound("rec-stop");
                LogToStderr("Audio files preloaded at startup");
            }
            catch (Exception ex)
            {
                LogToStderr($"Failed to initialize audio service: {ex.Message}");
            }
        }

        private void PreloadSound(string soundName)
        {
            try
            {
                var assembly = Assembly.GetExecutingAssembly();
                var resourceName = $"WindowsHelper.Resources.{soundName}.mp3";

                using var stream = assembly.GetManifestResourceStream(resourceName);
                if (stream != null)
                {
                    using var ms = new MemoryStream();
                    stream.CopyTo(ms);
                    preloadedAudio[soundName] = ms.ToArray();
                    LogToStderr($"Preloaded {soundName}.mp3 ({preloadedAudio[soundName].Length} bytes)");
                }
                else
                {
                    LogToStderr($"Resource not found for preloading: {resourceName}");
                }
            }
            catch (Exception ex)
            {
                LogToStderr($"Failed to preload {soundName}: {ex.Message}");
            }
        }

        public async Task PlaySound(string soundName, string requestId)
        {
            try
            {
                LogToStderr($"PlaySound called with soundName: {soundName}");

                // Stop any currently playing sound
                if (waveOut != null && waveOut.PlaybackState == PlaybackState.Playing)
                {
                    waveOut.Stop();
                    waveOut.Dispose();
                    waveOut = null;
                }

                // Use preloaded audio data (fast) or fall back to loading from resources
                byte[]? audioData;
                if (!preloadedAudio.TryGetValue(soundName, out audioData))
                {
                    LogToStderr($"Audio not preloaded, loading from resources: {soundName}");
                    var assembly = Assembly.GetExecutingAssembly();
                    var resourceName = $"WindowsHelper.Resources.{soundName}.mp3";

                    using var stream = assembly.GetManifestResourceStream(resourceName);
                    if (stream == null)
                    {
                        LogToStderr($"Resource not found: {resourceName}");
                        return;
                    }

                    using var ms = new MemoryStream();
                    await stream.CopyToAsync(ms);
                    audioData = ms.ToArray();
                }

                // Create memory stream from preloaded/loaded audio data
                using var memoryStream = new MemoryStream(audioData);
                using var audioFile = new Mp3FileReader(memoryStream);

                waveOut = new WaveOutEvent();
                waveOut.Init(audioFile);

                // Set up completion handler
                var completionSource = new TaskCompletionSource<bool>();
                waveOut.PlaybackStopped += (sender, args) =>
                {
                    LogToStderr($"Sound playback finished for {soundName}");
                    completionSource.TrySetResult(true);
                    SoundPlaybackCompleted?.Invoke(this, requestId);
                };

                // Start playback
                waveOut.Play();
                LogToStderr($"Playing sound: {soundName}.mp3");

                // Wait for completion
                await completionSource.Task;
            }
            catch (Exception ex)
            {
                LogToStderr($"Error playing sound {soundName}: {ex.Message}");
            }
            finally
            {
                if (waveOut != null)
                {
                    waveOut.Dispose();
                    waveOut = null;
                }
            }
        }

        public bool MuteSystemAudio()
        {
            try
            {
                LogToStderr("MuteSystemAudio called");
                
                if (deviceEnumerator == null) return false;
                
                var device = deviceEnumerator.GetDefaultAudioEndpoint(DataFlow.Render, Role.Multimedia);
                if (device != null)
                {
                    // Store original state
                    originalVolume = device.AudioEndpointVolume.MasterVolumeLevelScalar;
                    originalMuteState = device.AudioEndpointVolume.Mute;
                    
                    // Mute the audio
                    device.AudioEndpointVolume.Mute = true;
                    
                    LogToStderr($"System audio muted. Original volume: {originalVolume}, Original mute: {originalMuteState}");
                    return true;
                }
            }
            catch (Exception ex)
            {
                LogToStderr($"Error muting system audio: {ex.Message}");
            }
            
            return false;
        }

        public bool RestoreSystemAudio()
        {
            try
            {
                LogToStderr("RestoreSystemAudio called");
                
                if (deviceEnumerator == null) return false;
                
                var device = deviceEnumerator.GetDefaultAudioEndpoint(DataFlow.Render, Role.Multimedia);
                if (device != null)
                {
                    // Restore original state
                    device.AudioEndpointVolume.Mute = originalMuteState;
                    device.AudioEndpointVolume.MasterVolumeLevelScalar = originalVolume;
                    
                    LogToStderr($"System audio restored. Volume: {originalVolume}, Mute: {originalMuteState}");
                    return true;
                }
            }
            catch (Exception ex)
            {
                LogToStderr($"Error restoring system audio: {ex.Message}");
            }
            
            return false;
        }

        private void LogToStderr(string message)
        {
            var timestamp = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss.fff");
            Console.Error.WriteLine($"[{timestamp}] [AudioService] {message}");
            Console.Error.Flush();
        }
    }
}

File: packages/native-helpers/swift-helper/Sources/SwiftHelper/AudioService.swift
import AVFoundation
import Foundation

class AudioService: NSObject, AVAudioPlayerDelegate {
    private var audioPlayer: AVAudioPlayer?
    private var audioCompletionHandler: (() -> Void)?
    private var preloadedAudio: [String: Data] = [:]
    private let dateFormatter: DateFormatter

    override init() {
        self.dateFormatter = DateFormatter()
        self.dateFormatter.dateFormat = "yyyy-MM-dd HH:mm:ss.SSS"
        super.init()
        preloadSounds()
    }

    private func preloadSounds() {
        // Preload audio files at startup for faster playback
        preloadedAudio["rec-start"] = Data(PackageResources.rec_start_mp3)
        logToStderr("[AudioService] Preloaded rec-start.mp3 (\(preloadedAudio["rec-start"]?.count ?? 0) bytes)")

        preloadedAudio["rec-stop"] = Data(PackageResources.rec_stop_mp3)
        logToStderr("[AudioService] Preloaded rec-stop.mp3 (\(preloadedAudio["rec-stop"]?.count ?? 0) bytes)")

        logToStderr("[AudioService] Audio files preloaded at startup")
    }

    func playSound(named soundName: String, completion: (() -> Void)? = nil) {
        logToStderr("[AudioService] playSound called with soundName: \(soundName)")

        // Stop any currently playing sound
        if audioPlayer?.isPlaying == true {
            logToStderr(
                "[AudioService] Sound '\(audioPlayer?.url?.lastPathComponent ?? "previous")' is playing. Stopping it."
            )
            audioPlayer?.delegate = nil
            audioPlayer?.stop()
        }
        audioPlayer = nil
        audioCompletionHandler = nil

        audioCompletionHandler = completion

        // Use preloaded audio data (fast) or fall back to loading from resources
        let soundData: Data
        if let preloaded = preloadedAudio[soundName] {
            logToStderr("[AudioService] Using preloaded audio for \(soundName).mp3 (\(preloaded.count) bytes)")
            soundData = preloaded
        } else {
            logToStderr("[AudioService] Audio not preloaded, loading from PackageResources: \(soundName)")
            switch soundName {
            case "rec-start":
                soundData = Data(PackageResources.rec_start_mp3)
            case "rec-stop":
                soundData = Data(PackageResources.rec_stop_mp3)
            default:
                logToStderr("[AudioService] Error: Unknown sound name '\(soundName)'. Completion will not be called.")
                audioCompletionHandler = nil
                return
            }
        }

        do {
            audioPlayer = try AVAudioPlayer(data: soundData)
            audioPlayer?.delegate = self

            if audioPlayer?.play() == true {
                logToStderr("[AudioService] Playing sound: \(soundName).mp3. Delegate will handle completion.")
            } else {
                logToStderr(
                    "[AudioService] Failed to start playing sound: \(soundName).mp3. Completion will not be called."
                )
                audioCompletionHandler = nil
            }
        } catch {
            logToStderr(
                "[AudioService] Error initializing AVAudioPlayer for \(soundName).mp3: \(error.localizedDescription). Completion will not be called."
            )
            audioCompletionHandler = nil
        }
    }

    // MARK: - AVAudioPlayerDelegate

    func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
        logToStderr(
            "[AudioService] Sound playback finished (player URL: \(player.url?.lastPathComponent ?? "unknown"), successfully: \(flag))."
        )

        let handlerToCall = audioCompletionHandler
        audioCompletionHandler = nil

        if flag {
            logToStderr("[AudioService] Sound finished successfully. Executing completion handler.")
            handlerToCall?()
        } else {
            logToStderr("[AudioService] Sound did not finish successfully. Not executing completion handler.")
        }
    }

    private func logToStderr(_ message: String) {
        let timestamp = dateFormatter.string(from: Date())
        let logMessage = "[\(timestamp)] \(message)\n"
        FileHandle.standardError.write(logMessage.data(using: .utf8)!)
    }
}
